cmd/kubeadm/app/cmd/phases/certs.go

```diff
@@ -75,83 +94,76 @@ func NewCmdCerts() *cobra.Command {
+func getCertsSubPhases() []workflow.Phase {
+	subPhases := []workflow.Phase{}
 
-	// Default values for the cobra help text
-	kubeadmscheme.Scheme.Default(cfg)
+	certTree, _ := certsphase.GetDefaultCertList().AsMap().CertTree()
 
-	var cfgPath string
+	for ca, certList := range certTree {
+		caPhase := makePhase(ca, runCAPhase(ca))
+		subPhases = append(subPhases, caPhase)
 
-	// Special case commands
-	// All runs CreatePKIAssets, which isn't a particular certificate
-	allCmd := &cobra.Command{
-		Use:     "all",
-		Short:   "Generates all PKI assets necessary to establish the control plane",
-		Long:    allCertsLongDesc,
-		Example: allCertsExample,
-		Run: func(cmd *cobra.Command, args []string) {
-			internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(cfgPath, cfg)
-			kubeadmutil.CheckErr(err)
+		for _, cert := range certList {
+			certPhase := makePhase(cert, runCertPhase(cert, ca))
+			subPhases = append(subPhases, certPhase)
+		}
 
-			err = certsphase.CreatePKIAssets(internalcfg)
-			kubeadmutil.CheckErr(err)
-		},
+		subPhases = append(subPhases, caPhase)
 	}
-	addFlags(allCmd, &cfgPath, cfg, true)
 
 	// SA creates the private/public key pair, which doesn't use x509 at all
-	saCmd := &cobra.Command{
-		Use:   "sa",
+	saPhase := workflow.Phase{
+		Name:  "sa",
 		Short: "Generates a private key for signing service account tokens along with its public key",
 		Long:  saKeyLongDesc,
-		Run: func(cmd *cobra.Command, args []string) {
-			internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(cfgPath, cfg)
-			kubeadmutil.CheckErr(err)
-
-			err = certsphase.CreateServiceAccountKeyAndPublicKeyFiles(internalcfg)
-			kubeadmutil.CheckErr(err)
-		},
+		Run:   runCertsSa,
 	}
-	addFlags(saCmd, &cfgPath, cfg, false)
 
-	// "renew" command
-	renewCmd := certscmdphase.NewCmdCertsRenewal()
+	subPhases = append(subPhases, saPhase)
 
-	subCmds := []*cobra.Command{allCmd, saCmd, renewCmd}
+	return subPhases
+}
 
-	certTree, err := certsphase.GetDefaultCertList().AsMap().CertTree()
-	kubeadmutil.CheckErr(err)
+func runCertsSa(c workflow.RunData) error {
+	data, ok := c.(certsData)
+	if !ok {
+		return errors.New("certs phase invoked with an invalid data struct")
+	}
 
-	for ca, certList := range certTree {
-		// Don't use pointers from for loops, they will be rewrittenb
-		caCmds := makeCommandsForCA(ca, certList, &cfgPath, cfg)
-		subCmds = append(subCmds, caCmds...)
+	cfg := data.Cfg()
+	cfg.CertificatesDir = data.CertificateWriteDir()
+	defer func() { cfg.CertificatesDir = data.CertificateDir() }()
+
+	if err := certsphase.CreateServiceAccountKeyAndPublicKeyFiles(cfg); err != nil {
+		return err
 	}
 
-	return subCmds
+	return nil
 }
 
-func makeCmd(certSpec *certsphase.KubeadmCert, cfgPath *string, cfg *kubeadmapiv1beta1.InitConfiguration) *cobra.Command {
-	cmd := &cobra.Command{
-		Use:   certSpec.Name,
+func makePhase(certSpec *certsphase.KubeadmCert, run func(c workflow.RunData) error) workflow.Phase {
```

Comment:
we are inconsistent in how we name "the functions that create a sub-phase".
also this function is going to be available package in cmd/kubeadm/app/cmd/phases/.

probably best to name it:
newCertSubPhase()

similar to:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/phases/kubelet.go#L106

Reply:
Updated